fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/BoundRecordQuery.java

```diff
@@ -0,0 +1,172 @@
+
+    public BoundRecordQuery(@Nonnull final RecordStoreState recordStoreState, @Nonnull final RecordQuery recordQuery) {
+        this.recordStoreState = recordStoreState;
+        this.recordQuery = recordQuery;
+        this.parametersSupplier = Suppliers.memoize(this::computeParameters);
+        this.hashCodeSupplier = Suppliers.memoize(this::computeHashCode);
+    }
+
+    @Nonnull
+    public RecordStoreState getRecordStoreState() {
+        return recordStoreState;
+    }
+
+    @Nonnull
+    public RecordQuery getRecordQuery() {
+        return recordQuery;
+    }
+
+    public Set<String> getParameters() {
+        return parametersSupplier.get();
+    }
+
+    @Override
+    public int hashCode() {
+        return hashCodeSupplier.get();
+    }
+
+    @SuppressWarnings("UnstableApiUsage")
+    private int computeHashCode() {
+        final HashFunction hashFunction = Hashing.murmur3_32();
+        final Hasher hasher = hashFunction.newHasher();
+        hasher.putInt(recordQuery.hashCode());
+        hasher.putInt(recordStoreState.getIndexStates().hashCode());
+
+        getParameters().forEach(parameter -> hasher.putString(parameter, Charset.defaultCharset()));
+
+        final QueryComponent filter = recordQuery.getFilter();
+        if (filter != null) {
+            pairwiseDependentComparisons(groupedComparisons(filter))
+                    .forEach(pair -> hasher.putBoolean(Objects.equals(pair.getLeft().getComparand(), pair.getRight().getComparand())));
+        }
+        return hasher.hash().asInt();
+    }
+
+    private Set<String> computeParameters() {
+        final QueryComponent filter = recordQuery.getFilter();
+        if (filter != null) {
+            final List<Pair<ComponentWithComparison, List<Comparisons.Comparison>>> componentsAndComparisons =
+                    groupedComparisons(filter)
+                            .collect(ImmutableList.toImmutableList());
+
+            return componentsAndComparisons
+                    .stream()
+                    .flatMap(pair -> pair.getRight().stream())
+                    .flatMap(comparison -> comparison instanceof Comparisons.ComparisonWithParameter ? Stream.of(((Comparisons.ComparisonWithParameter)comparison).getParameter()) : Stream.empty())
+                    .collect(ImmutableSet.toImmutableSet());
+        }
+        return ImmutableSet.of();
+    }
+
+    @Override
+    @SuppressWarnings("UnstableApiUsage")
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof BoundRecordQuery)) {
+            return false;
+        }
+        final BoundRecordQuery other = (BoundRecordQuery)o;
+
+        if (!getRecordQuery().equals(other.getRecordQuery())) {
+            return false;
+        }
+
+        if (!getParameters().equals(other.getParameters())) {
+            return false;
+        }
+
+        if (!recordStoreState.compatibleWith(other.getRecordStoreState())) {
```

Comment:
Looking at `RecordStoreState.compatibleWith()`, I also wonder if it should take into account metadata version?